src/messages/HeadersMessage.cpp

```diff
@@ -0,0 +1,145 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "HeadersMessage.h"
+
+namespace {
+    bool HeadersMessage::handle(CNode *const pfrom,
+                                CDataStream &vRecv,
+                                int64_t nTimeReceived,
+                                std::string &strCommand,
+                                const bool xthinEnabled,
+                                const bool fReindexing) {
+        std::vector<CBlockHeader> headers;
+
+        // Bypass the normal CBlock deserialization, as we don't want to risk deserializing 2000 full blocks.
+        unsigned int nCount = ReadCompactSize(vRecv);
+        if (nCount > MAX_HEADERS_RESULTS) {
+            Misbehaving(pfrom->GetId(), 20);
+            return error("headers message size = %u", nCount);
+        }
+        headers.resize(nCount);
+        for (unsigned int n = 0; n < nCount; n++) {
+            vRecv >> headers[n];
+            ReadCompactSize(vRecv); // ignore tx count; assume it is 0.
+        }
+
+        LOCK(cs_main);
+
+        if (nCount == 0) {
+            // Nothing interesting. Stop asking this peers for more headers.
+            return true;
+        }
+
+        const CChainParams chainparams = Params();
+        CBlockIndex *pindexLast = NULL;
+        BOOST_FOREACH(const CBlockHeader &header, headers) {
+                        CValidationState state;
+                        if (pindexLast != NULL && header.hashPrevBlock != pindexLast->GetBlockHash()) {
+                            Misbehaving(pfrom->GetId(), 20);
+                            logWarning(Log::Net)
+                                    << "p2p HEADERS command received non-continues headers sequence from"
+                                    << pfrom->GetId();
+                            return false;
+                        }
+                        if (!AcceptBlockHeader(header, state, chainparams, &pindexLast)) {
+                            int nDoS;
+                            if (state.IsInvalid(nDoS)) {
+                                if (nDoS > 0)
+                                    Misbehaving(pfrom->GetId(), nDoS);
+                                logWarning(Log::Net) << "p2p HEADERS command received invalid header from"
+                                                     << pfrom->GetId();
+                                return false;
+                            }
+                        }
+                    }
+
+        if (pindexLast)
+            UpdateBlockAvailability(pfrom->GetId(), pindexLast->GetBlockHash());
```

Comment:
You forgot to add a single line to the main.h making this main.cpp method available to the entire application.

Reply:
ha, I had it and took it out because I was getting an unused warning